runtime/app.go

```diff
@@ -0,0 +1,152 @@
+			return fmt.Errorf("AppModule named %q already exists", name)
+		}
+		a.ModuleManager.Modules[name] = appModule
+
+		if _, ok := a.basicManager[name]; ok {
+			return fmt.Errorf("AppModuleBasic named %q already exists", name)
+		}
+
+		a.basicManager[name] = appModule
+		appModule.RegisterInterfaces(a.interfaceRegistry)
+		appModule.RegisterLegacyAminoCodec(a.amino)
+
+	}
+	return nil
+}
+
+// Load finishes all initialization operations and loads the app.
+func (a *App) Load(loadLatest bool) error {
+	if a.msgServiceRegistrar != nil {
+		configurator := module.NewConfigurator(a.cdc, a.msgServiceRegistrar, a.GRPCQueryRouter())
+		a.ModuleManager.RegisterServices(configurator)
+	}
+
+	if len(a.config.InitGenesis) != 0 {
+		a.ModuleManager.SetOrderInitGenesis(a.config.InitGenesis...)
+		a.SetInitChainer(a.InitChainer)
+	}
+
+	if len(a.config.BeginBlockers) != 0 {
+		a.ModuleManager.SetOrderBeginBlockers(a.config.BeginBlockers...)
+		a.SetBeginBlocker(a.BeginBlocker)
+	}
+
+	if len(a.config.EndBlockers) != 0 {
+		a.ModuleManager.SetOrderEndBlockers(a.config.EndBlockers...)
+		a.SetEndBlocker(a.EndBlocker)
+	}
+
+	if loadLatest {
+		if err := a.LoadLatestVersion(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// BeginBlocker application updates every begin block
+func (app *App) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) abci.ResponseBeginBlock {
+	return app.ModuleManager.BeginBlock(ctx, req)
+}
+
+// EndBlocker application updates every end block
+func (app *App) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
+	return app.ModuleManager.EndBlock(ctx, req)
+}
+
+// InitChainer initializes the chain.
+func (a *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+	var genesisState map[string]json.RawMessage
+	if err := json.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
+		panic(err)
+	}
+	return a.ModuleManager.InitGenesis(ctx, a.cdc, genesisState)
+}
+
+// RegisterAPIRoutes registers all application module routes with the provided
+// API server.
+func (a *App) RegisterAPIRoutes(apiSvr *api.Server, _ config.APIConfig) {
+	clientCtx := apiSvr.ClientCtx
+	// Register new tx routes from grpc-gateway.
+	authtx.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
+	// Register new tendermint queries routes from grpc-gateway.
+	tmservice.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
+
+	// Register grpc-gateway routes for all modules.
+	a.basicManager.RegisterGRPCGatewayRoutes(clientCtx, apiSvr.GRPCGatewayRouter)
+}
+
+// RegisterTxService implements the Application.RegisterTxService method.
+func (a *App) RegisterTxService(clientCtx client.Context) {
+	authtx.RegisterTxService(a.GRPCQueryRouter(), clientCtx, a.Simulate, a.interfaceRegistry)
+}
+
+// RegisterTendermintService implements the Application.RegisterTendermintService method.
+func (a *App) RegisterTendermintService(clientCtx client.Context) {
```

Comment:
these 2 seem weird in the runtime. Can they be modules?

Reply:
yes, they should be. at least they shouldn't be standalone methods on `server.Application`. maybe runtime should register the tendermint service, but authtx should either go in auth or we should have a standalone `tx` or `auth/tx` module. should we add an issue as part of the epic to fix this?